### PR TITLE
8297933: [REDO] Compiler should only use verified interface types for optimization

### PR DESCRIPTION
**Pull request by rwestrel**
This PR re-does 6312651 (Compiler should only use verified interface
types for optimization) with a couple fixes I had pushed afterward
(8297556 and 8297343) and fixes for some other issues.

The trickiest one is a fix for 8297345 (C2: SIGSEGV in
PhaseIdealLoop::push_pinned_nodes_thru_region) for which I added a
test case. The crash occurs because a (If (Bool (CmpP (LoadKlass ..))))
only has a single projection. It lost the other projection because of
a CheckCastPP that becomes top. Initially the pattern is, in pseudo
code,:

if (obj.klass == some_class) {
  obj = CheckCastPP#1(obj);
}

obj itself is a CheckCastPP that's pinned at a dominating if. That
dominating if goes through split through phi. The LoadKlass for the
pseudo code above also has control set to the dominating if being
transformed. This result in:

if (phi1 == some_class) {
  obj = CheckCastPP#1(phi2);
}

with phi1 = (Phi (LoadKlass obj) (LoadKlass obj)) and phi2 = (Phi obj obj)
with obj = (CheckCastPP#2 obj')

PhiNode::Ideal() transforms phi2 into a new CheckCastPP:
(CheckCastPP#3 obj' obj') with control set to the region right above
the if in the pseudo code above. There happens to be another
CheckCastPP at the same control which casts obj' to a narrower
type. So the new CheckCastPP#3 is replaced by that one (because of
ConstraintCastNode::dominating_cast())and pseudo code becomes:

if (phi1 == some_class) {
  obj = CheckCastPP#1(CheckCastPP#4(obj'));
}

and then:

if (phi1 == some_class) {
  obj = top;
}

because the types of the 2 CheckCastPPs conflict. That would be ok if:

phi1 == some_class

would constant fold. It would if the test was:

if (CheckCastPP#4(obj').klass == some_klass) {

but because of split if, the (CmpP (LoadKlass ..)) and the
CheckCastPP#1 ended up with 2 different object inputs that then were
transformed differently. The fix I propose is to have split if clone the entire:

(Bool (CmpP (LoadKlass (AddP ..))))

down the same way (Bool (CmpP ..)) is cloned down. After split if, the
pseudo code becomes:

if (phi.klass == some_class) {
  obj = CheckCastPP#1(phi);
}

The bug can't occur because the CheckCastPP and (CmpP (LoadKlass ..))
operate on the same phi input. The change in split_if.cpp implements
that.

The other fixes are:

- arraycopynode.cpp: a crash happens because dest_offset and
  src_offset are the same. The call to transform that results in
  src_scale, causes src_offset (and thus dest_offset) to become
  dead. The fix is to add a hook node to preserve dest_offset. This is
  unrelated to 6312651 but it triggers with that change for some
  reason.
  
- castnode.cpp: I removed CheckCastPPNode::Identity(), a piece of code
  that the change in the handling of interfaces make obsolete and that
  I missed in the PR for 6312651.
  
- castnode.cpp: the change in CheckCastPPNode::Value() fixes a rare
  assert when during CCP, Value() is called with an input raw constant
  ptr.
  
- type.cpp: a _klass = NULL field in arrays used to indicate only top
  or bottom but I changed that so _klass is only guaranteed non null
  for basic type arrays. The fix in type.cpp updates a piece of code
  that I didn't adapt to the new meaning of _klass = NULL.
  
- the other changes are due to StressReflectiveCode. With 6312651, a
  CheckCastPP can fold to top if it sees a type for its input that
  conflicts with its own type. That wasn't the case before. So if a
  type check fails, a CheckCastPP will fold to top and the control
  flow branch it's in must die. That doesn't always happen with
  StressReflectiveCode: the CheckCastPP folds but not the control flow
  path. With ExpandSubTypeCheckAtParseTime on, that's because of a
  code path in LoadNode::Value() that's disabled with
  StressReflectiveCode. With ExpandSubTypeCheckAtParseTime off, it's
  because Compile::static_subtype_check() is always pessimistic with
  StressReflectiveCode but it's used by SubTypeCheckNode::sub() to
  find when a node can constant fold.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297933](https://bugs.openjdk.org/browse/JDK-8297933): [REDO] Compiler should only use verified interface types for optimization

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [848fc8df](https://git.openjdk.org/jdk/pull/11666/files/848fc8dfaf64e060cfb0c2d286824f2467c6fd40)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11666/head:pull/11666` \
`$ git checkout pull/11666`

Update a local copy of the PR: \
`$ git checkout pull/11666` \
`$ git pull https://git.openjdk.org/jdk pull/11666/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11666`

View PR using the GUI difftool: \
`$ git pr show -t 11666`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11666.diff">https://git.openjdk.org/jdk/pull/11666.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+ */
+
+/**
+ * @test
+ * @bug 8297345
+ * @summary C2: SIGSEGV in PhaseIdealLoop::push_pinned_nodes_thru_region
+ *
+ * @run main/othervm -XX:-TieredCompilation -XX:-UseOnStackReplacement -XX:-BackgroundCompilation
+ *                   -XX:CompileOnly=TestCheckCastPPBecomesTOP::test1 -XX:LoopMaxUnroll=0
+ *                   -XX:CompileCommand=dontinline,TestCheckCastPPBecomesTOP::notInlined -XX:+UseParallelGC TestCheckCastPPBecomesTOP
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:31]: This test fails when executed with a different GC (you need to remove the explicit setting or add an `@requires`:
```
Error occurred during initialization of VM
Multiple garbage collectors selected
```

(I did some more testing in top of what I already did last week - that's why it didn't show up earlier)

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:32]: I added a `@requires`

File: test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
  * @test
  * @bug 8297345
  * @summary C2: SIGSEGV in PhaseIdealLoop::push_pinned_nodes_thru_region
+ * @requires vm.gc.Parallel
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: I think it should be:

```suggestion
 * @requires vm.gc == null | vm.gc.Parallel
```

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: FTR, I discussed this privately with Tobias. His concern is that the test would not be run if not passed `-XX:+UseParallelGC`. That doesn't seem to be an issue though as the test runs as expected when passed no option or `-XX:+UseParallelGC` and doesn't run with a command line option that enables some other gc.

Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: Which means it will run with default GC: G1.

**Review by vnkozlov — COMMENTED** [review]
@rwestrel Can you push split-if/CheckCastPP fix separately, before these changes?
It is not directly related to this work.

**Comment by rwestrel** [comment]
> @rwestrel Can you push split-if/CheckCastPP fix separately, before these changes? It is not directly related to this work.

The split if issue is not there without the rest of the change. Do you want it as a seperate change anyway?

File: test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java (hunk 2 above)
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: Which is the correct way? `@requires vm.gc == null | vm.gc.Parallel` or `@requires vm.gc.Parallel`?

**Comment by vnkozlov** [comment]
> > @rwestrel Can you push split-if/CheckCastPP fix separately, before these changes? It is not directly related to this work.
> 
> The split if issue is not there without the rest of the change. Do you want it as a separate change anyway?

I understand that. But it is significant change and I want to push it first and test separately.

File: test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java (hunk 2 above)
Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: `@requires vm.gc.Parallel` 

 [JDK-8160088](https://bugs.openjdk.org/browse/JDK-8160088)  changed it from original `@requires vm.gc == "null" | vm.gc == "Parallel"`

Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:28]: Clarifying. `vm.gc.Parallel` is `true` when ParallelGC supported and its flag is specified on command line or no GC flags specified and GC is selected ergonomically. Which means the same as original `@requires`

**Comment by rwestrel** [comment]
> I understand that. But it is significant change and I want to push it first and test separately.

I filed JDK-8298848 for that.

**Comment by rwestrel** [comment]
> > I understand that. But it is significant change and I want to push it first and test separately.
> 
> I filed JDK-8298848 for that.

And I removed the splif if change from this PR.

File: src/hotspot/share/ci/ciInstanceKlass.cpp [hunk 3]
```diff
@@ -729,6 +730,31 @@ void ciInstanceKlass::dump_replay_instanceKlass(outputStream* out, InstanceKlass
...
+  }
+  return _transitive_interfaces;
+}
+
+void ciInstanceKlass::compute_transitive_interfaces() {
+  GUARDED_VM_ENTRY(
+          InstanceKlass* ik = get_instanceKlass();
+          Array<InstanceKlass*>* interfaces = ik->transitive_interfaces();
+          Arena* arena = CURRENT_ENV->arena();
+          int len = interfaces->length() + (is_interface() ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (rwestrel) [reply src/hotspot/share/ci/ciInstanceKlass.cpp]: Done in new commit.

**Review by vnkozlov — COMMENTED** [review]
Looks good to me. Lets finish it after #11689 is pushed.

File: src/hotspot/share/ci/ciArrayKlass.hpp [hunk 4]
```diff
@@ -56,6 +56,7 @@ class ciArrayKlass : public ciKlass {
   bool is_array_klass() const { return true; }
 
   static ciArrayKlass* make(ciType* element_type);
+
```
Comment (iwanowww) [inline src/hotspot/share/ci/ciArrayKlass.hpp:59]: Redundant. A leftover from reverted changes?

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/share/ci/ciArrayKlass.hpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/ci/ciArrayKlass.hpp]: Removed.

**Comment by rwestrel** [comment]
Hi @vnkozlov 

> Looks good to me. Lets finish it after #11689 is pushed.

Can I push it now or do you want to take another look? Thanks.

**Comment by chhagedorn** [comment]
I'll submit some testing again with the latest state over the weekend. Will report back on Monday.

**Comment by vnkozlov** [comment]
> Hi @vnkozlov
> 
> > Looks good to me. Lets finish it after #11689 is pushed.
> 
> Can I push it now or do you want to take another look? Thanks.

Lets wait testing results from Christian. Changes looks fine to me now.

**Comment by chhagedorn** [comment]
All tests passed!

**Comment by rwestrel** [comment]
@vnkozlov @iwanowww thanks for the reviews, @chhagedorn thanks for running tests

**Pull request closed without merging**